WalletWasabi/Blockchain/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -257,6 +258,27 @@ private ProcessedResult ProcessNoLock(SmartTransaction tx)
 			result.SpentCoins.Add(coin);
 			Coins.Spend(coin, tx);
 
+			// If the key is internal and doesn't contain any other coin, then it won't be used anymore.
+			if (tx.Confirmed && KeyManager.TryGetKeyForScriptPubKey(coin.ScriptPubKey, out HdPubKey? spenderKey))
+			{
+				if (spenderKey.IsInternal)
+				{
+					if (spenderKey.Coins.Except(result.SpentCoins).All(x => x.Outpoint == coin.Outpoint))
+					{
+						KeyManager.SetKeyState(KeyState.Obsolete, spenderKey);
+						if (spenderKey.ObsoleteHeight == 0 || spenderKey.ObsoleteHeight < tx.Height)
+						{
+							spenderKey.ObsoleteHeight = tx.Height;
```

Comment:
There are two cases here:

- One, the key was not obsolete and now it will become. 
- Second, the key was already obsolete but now a new spender tx was found (luckily), and in that update the `ObsoleteHeight`.